sel/src/Sel/Internal.hs

```diff
@@ -1,53 +1,103 @@
--- libsodium's @memcmp@ in cases when the return code is necessary.
-foreign import capi unsafe "string.h memcmp"
-  memcmp :: Ptr a -> Ptr b -> CSize -> IO CInt
+-- | Compare the contents of two byte arrays in constant time.
+--
+-- /See:/ [Constant-time test for equality](https://doc.libsodium.org/helpers#constant-time-test-for-equality)
+--
+-- @since 0.0.3.0
+foreignPtrEqConstantTime :: ForeignPtr CUChar -> ForeignPtr CUChar -> CSize -> IO Bool
+foreignPtrEqConstantTime p q size =
+  fmap (== 0) . use $
+    sodiumMemcmp <$> foreignPtr p <*> foreignPtr q <*> pure size
 
--- | Compare if the contents of two @ForeignPtr@s are equal.
-foreignPtrEq :: ForeignPtr a -> ForeignPtr a -> CSize -> IO Bool
-foreignPtrEq fptr1 fptr2 size =
-  withForeignPtr fptr1 $ \p ->
-    withForeignPtr fptr2 $ \q ->
-      do
-        result <- memcmp p q size
-        return $ 0 == result
+-- | Lexicographically compare the contents of two byte arrays.
+--
+-- ⚠️ Such comparisons are vulnerable to timing attacks, and should be
+-- avoided for secret data.
+--
+-- @since 0.0.1.0
+foreignPtrOrd :: ForeignPtr CUChar -> ForeignPtr CUChar -> CSize -> IO Ordering
+foreignPtrOrd p q size =
+  fmap (`compare` 0) . useM $
+    memcmp
+      <$> foreignPtr (coerce p)
+      <*> foreignPtr (coerce q)
+      <*> pure (fromIntegral size)
 
--- | Compare the contents of two @ForeignPtr@s using lexicographical ordering.
-foreignPtrOrd :: ForeignPtr a -> ForeignPtr a -> CSize -> IO Ordering
-foreignPtrOrd fptr1 fptr2 size =
-  withForeignPtr fptr1 $ \p ->
-    withForeignPtr fptr2 $ \q ->
-      do
-        result <- memcmp p q size
-        return $
-          if
-            | result == 0 -> EQ
-            | result < 0 -> LT
-            | otherwise -> GT
+-- | Compare two byte arrays for lexicographic equality.
+--
+-- ⚠️ Such comparisons are vulnerable to timing attacks, and should be
+-- avoided for secret data.
+--
+-- @since 0.0.1.0
+foreignPtrEq :: ForeignPtr CUChar -> ForeignPtr CUChar -> CSize -> IO Bool
+foreignPtrEq p q size = (== EQ) <$> foreignPtrOrd p q size
 
+-- | Convert a @'ForeignPtr' a@ to a 'ByteString' of the given length
+-- and render the hexadecimal-encoded bytes as a 'String'.
+--
+-- @since 0.0.1.0
 foreignPtrShow :: ForeignPtr a -> CSize -> String
-foreignPtrShow fptr size =
-  BS.unpackChars . Base16.extractBase16 . Base16.encodeBase16' $
-    BS.fromForeignPtr (Foreign.castForeignPtr fptr) 0 (fromIntegral @CSize @Int size)
+foreignPtrShow (Foreign.castForeignPtr -> cstring) size =
+  ByteString.unpackChars . Base16.extractBase16 . Base16.encodeBase16' $
+    ByteString.fromForeignPtr cstring 0 (fromIntegral size)
+
+-- | Copy a byte array as key material.
+--
+-- The size of the array is not checked. The input may be truncated if
+-- it is too long, or an unchecked exception may be thrown if it is
+-- too short.
+--
+-- @since 0.0.3.0
+unsafeCopyToSodiumPointer :: CSize -> StrictByteString -> IO (ForeignPtr CUChar)
```

Comment:
This isn't used in this PR, but is used in the next PR in the stack: https://github.com/jhenahan/libsodium-bindings/pull/2/files#diff-6d90c4955b4c0bcc7d190b7109069933f66133e3a46f1fa0a64e6308f33f4707R499